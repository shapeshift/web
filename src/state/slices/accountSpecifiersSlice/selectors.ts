import type { ChainId } from '@shapeshiftoss/caip'
import { createSelector } from 'reselect'
import type { ReduxState } from 'state/reducer'

import { createDeepEqualOutputSelector } from './../../selector-utils'

export const selectAccountSpecifiers = createDeepEqualOutputSelector(
  (state: ReduxState) => state.accountSpecifiers.accountSpecifiers,
  accountSpecifiers => accountSpecifiers,
)

// returns an array of the full `chainId:pubkeyish` type, as used in URLs for account pages
export const selectAccountSpecifierStrings = (state: ReduxState) =>
  state.accountSpecifiers.accountSpecifiers.map(accountSpecifier =>
    Object.entries(accountSpecifier)[0].join(':'),
  )

// Returns a ChainId-indexed object with all the `chainId:pubkeyish` accounts for that chainId
// For most accounts, that's effectively an AssetId, but not for e.g UTXO chains thus the pubkeyish naming
// We use this in cosmos plugin to get the pubkey as an AssetId, without needing to use chain-adapters in components
// Since the pubkey is already in state
export const selectAccountSpecifiersByChainId = (state: ReduxState, chainId: ChainId) =>
  state.accountSpecifiers.accountSpecifiers.reduce<string[]>((acc, accountSpecifier) => {
    const pubkeyish = Object.entries(accountSpecifier)[0].join(':')
    const currentChainId = Object.keys(accountSpecifier)[0]

    if (currentChainId !== chainId) return acc

    acc.push(pubkeyish)

    return acc
  }, [])

export const selectFirstAccountSpecifierByChainId = createSelector(
  selectAccountSpecifiersByChainId,
  accountSpecifiers => accountSpecifiers[0],
)
