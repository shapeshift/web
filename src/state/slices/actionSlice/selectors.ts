import { selectEnabledWalletAccountIds } from '../common-selectors'
import { swapSlice } from '../swapSlice/swapSlice'
import { actionSlice } from './actionSlice'
import type { LimitOrderAction, RfoxClaimAction, TcyClaimAction } from './types'
import {
  ActionStatus,
  ActionType,
  isGenericTransactionAction,
  isLimitOrderAction,
  isPendingSwapAction,
  isRfoxClaimAction,
  isSwapAction,
  isTcyClaimAction,
} from './types'

import { createDeepEqualOutputSelector } from '@/state/selector-utils'
import {
  selectCowSwapQuoteIdParamFromRequiredFilter,
  selectSwapIdParamFromFilter,
} from '@/state/selectors'

export const selectActions = createDeepEqualOutputSelector(
  actionSlice.selectors.selectActionsById,
  actionSlice.selectors.selectActionIds,
  (actionsById, actionIds) => actionIds.map(id => actionsById[id]),
)

export const selectWalletActions = createDeepEqualOutputSelector(
  selectActions,
  selectEnabledWalletAccountIds,
  swapSlice.selectors.selectSwapsById,
  (actions, enabledWalletAccountIds, swapsById) => {
    return actions.filter(action => {
      if (isSwapAction(action)) {
        const swapId = action.swapMetadata.swapId
        const relatedSwap = swapsById[swapId]

        if (!relatedSwap?.sellAccountId) return false

        return enabledWalletAccountIds.includes(relatedSwap.sellAccountId)
      }

      if (isLimitOrderAction(action)) {
        return enabledWalletAccountIds.includes(action.limitOrderMetadata.accountId)
      }

      if (isGenericTransactionAction(action)) {
        return enabledWalletAccountIds.includes(action.transactionMetadata.accountId)
      }

      if (isRfoxClaimAction(action)) {
        return enabledWalletAccountIds.includes(
          action.rfoxClaimActionMetadata.request.stakingAssetAccountId,
        )
      }

      if (isTcyClaimAction(action)) {
        return enabledWalletAccountIds.includes(action.tcyClaimActionMetadata.claim.accountId)
      }

      return action
    })
  },
)

export const selectWalletActionsSorted = createDeepEqualOutputSelector(
  selectWalletActions,
  actions => {
    return actions
      .filter(action => action.status !== ActionStatus.Idle)
      .sort((a, b) => b.updatedAt - a.updatedAt)
  },
)

export const selectWalletPendingActions = createDeepEqualOutputSelector(
  selectWalletActions,
  actions => {
    return actions.filter(action => action.status === ActionStatus.Pending)
  },
)

export const selectPendingSwapActions = createDeepEqualOutputSelector(
  selectWalletActions,
  actions => {
    return actions.filter(isPendingSwapAction)
  },
)

export const selectSwapActionBySwapId = createDeepEqualOutputSelector(
  actionSlice.selectors.selectActionsById,
  actionSlice.selectors.selectActionIds,
  selectSwapIdParamFromFilter,
  (actionsById, actionIds, swapId) => {
    const actionId = actionIds.find(id => {
      const action = actionsById[id]
      return isSwapAction(action) && action.swapMetadata.swapId === swapId
    })

    if (!actionId) return

    return actionsById[actionId]
  },
)

export const selectOpenLimitOrderActionsFilteredByWallet = createDeepEqualOutputSelector(
  selectActions,
  selectEnabledWalletAccountIds,
  (actions, enabledWalletAccountIds) => {
    return actions.filter(
      action =>
        action.status === ActionStatus.Open &&
        action.type === ActionType.LimitOrder &&
        ((action.limitOrderMetadata.accountId &&
          enabledWalletAccountIds.includes(action.limitOrderMetadata.accountId)) ||
          !action.limitOrderMetadata.accountId),
    )
  },
)

export const selectLimitOrderActionByCowSwapQuoteId = createDeepEqualOutputSelector(
  selectActions,
  selectCowSwapQuoteIdParamFromRequiredFilter,
  (actions, cowSwapQuoteId) => {
    return actions.find(
      action =>
        isLimitOrderAction(action) && action.limitOrderMetadata.cowSwapQuoteId === cowSwapQuoteId,
    )
  },
)

export const selectLimitOrderActionsByWallet = createDeepEqualOutputSelector(
  selectActions,
  selectEnabledWalletAccountIds,
  (actions, enabledWalletAccountIds) => {
    return actions.filter(
      (action): action is LimitOrderAction =>
        isLimitOrderAction(action) &&
        enabledWalletAccountIds.includes(action.limitOrderMetadata.accountId),
    )
  },
)

export const selectRfoxClaimActionsByWallet = createDeepEqualOutputSelector(
  selectActions,
  selectEnabledWalletAccountIds,
  (actions, enabledWalletAccountIds) => {
    return actions.filter(
      (action): action is RfoxClaimAction =>
        isRfoxClaimAction(action) &&
        enabledWalletAccountIds.includes(
          action.rfoxClaimActionMetadata.request.stakingAssetAccountId,
        ),
    )
  },
)

export const selectPendingRfoxClaimActions = createDeepEqualOutputSelector(
  selectRfoxClaimActionsByWallet,
  actions => {
    return actions.filter(action => action.status === ActionStatus.Pending)
  },
)

export const selectTcyClaimActionsByWallet = createDeepEqualOutputSelector(
  selectActions,
  selectEnabledWalletAccountIds,
  (actions, enabledWalletAccountIds) => {
    return actions.filter(
      (action): action is TcyClaimAction =>
        isTcyClaimAction(action) &&
        enabledWalletAccountIds.includes(action.tcyClaimActionMetadata.claim.accountId),
    )
  },
)

export const selectPendingTcyClaimActions = createDeepEqualOutputSelector(
  selectTcyClaimActionsByWallet,
  actions => {
    return actions.filter(action => action.status === ActionStatus.Pending)
  },
)
