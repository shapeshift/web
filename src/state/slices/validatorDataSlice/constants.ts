export const SHAPESHIFT_COSMOS_VALIDATOR_ADDRESS =
  'cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf'
export const SHAPESHIFT_OSMOSIS_VALIDATOR_ADDRESS =
  'osmovaloper1vkfmegxrsveefn2wmudh7apxuzu2n77654ad62'
