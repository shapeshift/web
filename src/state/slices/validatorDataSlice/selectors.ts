import { createSelector } from '@reduxjs/toolkit'
import type { cosmos } from '@shapeshiftoss/chain-adapters'
import type { ReduxState } from 'state/reducer'

import type { PubKey, ValidatorData, ValidatorDataByPubKey } from './validatorDataSlice'

export const selectValidatorAddress = (_state: ReduxState, validatorAddress: PubKey) =>
  validatorAddress

export const selectValidatorData = (state: ReduxState): ValidatorData => state.validatorData
export const selectValidators = createSelector(
  selectValidatorData,
  (validatorData): ValidatorDataByPubKey => validatorData.byValidator,
)

export const selectValidatorByAddress = createSelector(
  selectValidatorData,
  selectValidatorAddress,
  (stakingData, validatorAddress): cosmos.Validator | null => {
    return stakingData.byValidator[validatorAddress] || null
  },
)
