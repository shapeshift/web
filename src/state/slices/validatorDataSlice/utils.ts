import type { AccountId, AssetId, ChainId } from '@shapeshiftoss/caip'
import { cosmosChainId, fromAccountId, fromAssetId, osmosisChainId } from '@shapeshiftoss/caip'
import flow from 'lodash/flow'

import {
  SHAPESHIFT_COSMOS_VALIDATOR_ADDRESS,
  SHAPESHIFT_OSMOSIS_VALIDATOR_ADDRESS,
} from './constants'

export const getDefaultValidatorAddressFromChainId = (chainId: ChainId) => {
  switch (chainId) {
    case cosmosChainId:
      return SHAPESHIFT_COSMOS_VALIDATOR_ADDRESS
    case osmosisChainId:
      return SHAPESHIFT_OSMOSIS_VALIDATOR_ADDRESS
    default:
      throw new Error(`chainId ${chainId} is not a valid Cosmos SDK chainId`)
  }
}
export const getDefaultValidatorAddressFromAssetId = flow([
  (assetId: AssetId) => fromAssetId(assetId).chainId,
  getDefaultValidatorAddressFromChainId,
])

export const getDefaultValidatorAddressFromAccountId = flow(
  (accountId: AccountId) => fromAccountId(accountId).chainId,
  getDefaultValidatorAddressFromChainId,
)
