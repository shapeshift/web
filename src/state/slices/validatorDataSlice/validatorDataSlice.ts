import { createSlice } from '@reduxjs/toolkit'
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { cosmosChainId } from '@shapeshiftoss/caip'
import { cosmossdk } from '@shapeshiftoss/chain-adapters'
// @ts-ignore this will fail at 'file differs in casing' error
import { chainAdapters } from '@shapeshiftoss/types'
import { getChainAdapters } from 'context/PluginProvider/PluginProvider'

export type PubKey = string

type SingleValidatorDataArgs = { validatorAddress: PubKey }

export type Validators = {
  validators: chainAdapters.cosmos.Validator[]
}

export type ValidatorData = {
  byValidator: ValidatorDataByPubKey
  validatorIds: PubKey[]
}

export type ValidatorDataByPubKey = {
  [k: PubKey]: chainAdapters.cosmos.Validator
}

const initialState: ValidatorData = {
  byValidator: {},
  validatorIds: [],
}

const updateOrInsertValidatorData = (
  validatorDataState: ValidatorData,
  validators: chainAdapters.cosmos.Validator[],
) => {
  validators.forEach(validator => {
    validatorDataState.validatorIds.push(validator.address)
    validatorDataState.byValidator[validator.address] = validator
  })
}

export const validatorData = createSlice({
  name: 'validatorData',
  initialState,
  reducers: {
    clear: () => initialState,
    upsertValidatorData: (
      validatorDataState,
      { payload }: { payload: { validators: chainAdapters.cosmos.Validator[] } },
    ) => {
      updateOrInsertValidatorData(validatorDataState, payload.validators)
    },
  },
})

export const validatorDataApi = createApi({
  reducerPath: 'validatorDataApi',
  // not actually used, only used to satisfy createApi, we use a custom queryFn
  baseQuery: fetchBaseQuery({ baseUrl: '/' }),
  // 5 minutes caching against overfetching. The only thing that can change on new Tx is effectively TVL and APR
  // The first won't noticeably change given the Million fiat precision we use, and the former effectively won't noticeably change either in such timeframe
  keepUnusedDataFor: 300,
  // refetch if network connection is dropped, useful for mobile
  refetchOnReconnect: true,
  endpoints: build => ({
    getValidatorData: build.query<chainAdapters.cosmos.Validator, SingleValidatorDataArgs>({
      queryFn: async ({ validatorAddress }, { dispatch }) => {
        const chainAdapters = getChainAdapters()
        const adapter = (await chainAdapters.byChainId(
          cosmosChainId,
        )) as cosmossdk.cosmos.ChainAdapter
        try {
          const data = await adapter.getValidator(validatorAddress)
          dispatch(
            validatorData.actions.upsertValidatorData({
              validators: [data],
            }),
          )
          return { data }
        } catch (e) {
          console.error('Error fetching single validator data', e)
          return {
            error: {
              data: `Error fetching validator data`,
              status: 500,
            },
          }
        }
      },
    }),
  }),
})
