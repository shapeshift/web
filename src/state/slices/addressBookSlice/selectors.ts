import { CHAIN_NAMESPACE, fromAccountId, fromChainId } from '@shapeshiftoss/caip'
import { matchSorter } from 'match-sorter'
import { createSelector } from 'reselect'

import { addressBookSlice } from './addressBookSlice'

import { findUtxoAccountIdByAddress } from '@/lib/utils/accounts'
import { createDeepEqualOutputSelector } from '@/state/selector-utils'
import {
  selectAccountAddressParamFromFilter,
  selectChainIdParamFromFilter,
  selectSearchQueryFromFilter,
} from '@/state/selectors'
import { selectAccountIdsWithoutEvms, selectEvmAccountIds } from '@/state/slices/common-selectors'

export const selectAddressBookEntries = createSelector(
  [addressBookSlice.selectors.selectEntriesByAccountId],
  byAccountId => Object.values(byAccountId),
)

export const selectAddressBookEntriesByChainId = createDeepEqualOutputSelector(
  selectChainIdParamFromFilter,
  selectAddressBookEntries,
  (chainId, entries) => {
    if (!chainId) return []

    return entries.filter(entry => entry.chainId === chainId)
  },
)

export const selectAddressBookEntriesBySearchQuery = createDeepEqualOutputSelector(
  selectChainIdParamFromFilter,
  selectSearchQueryFromFilter,
  selectAddressBookEntriesByChainId,
  (_, searchQuery, entries) => {
    if (!searchQuery) return entries

    const matchedEntries = matchSorter(entries, searchQuery, {
      keys: [
        { key: 'label', threshold: matchSorter.rankings.MATCHES },
        { key: 'address', threshold: matchSorter.rankings.CONTAINS },
      ],
    })

    return matchedEntries
  },
)

/**
 * Checks if a given address exists in the address book
 * Handles EVM namespace matching (same address across all EVM chains)
 */
export const selectIsAddressInAddressBook = createDeepEqualOutputSelector(
  selectAccountAddressParamFromFilter,
  selectChainIdParamFromFilter,
  selectAddressBookEntries,
  (accountAddress, chainId, entries) => {
    if (!accountAddress || !chainId) return false

    const normalizedAddress = accountAddress.toLowerCase()
    const { chainNamespace } = fromChainId(chainId)

    // For EVM chains, check if address exists in any EVM entry
    if (chainNamespace === CHAIN_NAMESPACE.Evm) {
      return entries.some(entry => {
        const entryChainNamespace = fromChainId(entry.chainId).chainNamespace
        return (
          entry.isExternal &&
          entryChainNamespace === CHAIN_NAMESPACE.Evm &&
          entry.address.toLowerCase() === normalizedAddress
        )
      })
    }

    // For non-EVM chains, check for exact chain + address match
    return entries.some(
      entry =>
        entry.isExternal &&
        entry.chainId === chainId &&
        entry.address.toLowerCase() === normalizedAddress,
    )
  },
)

/**
 * Checks if a given address belongs to one of the user's connected accounts
 * Handles EVM namespace matching (same address across all EVM chains)
 * For UTXO accounts, uses cached addresses from unchained
 * Returns the AccountId if found, null otherwise
 */
export const selectInternalAccountIdByAddress = createDeepEqualOutputSelector(
  selectAccountAddressParamFromFilter,
  selectChainIdParamFromFilter,
  selectAccountIdsWithoutEvms,
  selectEvmAccountIds,
  (accountAddress, chainId, accountIdsWithoutEvms, evmAccountIds) => {
    if (!accountAddress || !chainId) return null

    const normalizedAddress = accountAddress.toLowerCase()
    const { chainNamespace } = fromChainId(chainId)

    // For EVM chains, find first matching EVM account
    if (chainNamespace === CHAIN_NAMESPACE.Evm) {
      const accountId = evmAccountIds.find(accountId => {
        const { account, chainId: accountChainId } = fromAccountId(accountId)
        const accountChainNamespace = fromChainId(accountChainId).chainNamespace

        return (
          accountChainNamespace === CHAIN_NAMESPACE.Evm &&
          account.toLowerCase() === normalizedAddress
        )
      })

      return accountId ?? null
    }

    // For UTXO chains, use progressive address derivation to find matches
    // This intelligently searches in batches (20, 50, 100) to handle edge cases
    // where users have many transactions and the address is beyond the standard gap limit
    if (chainNamespace === CHAIN_NAMESPACE.Utxo) {
      return findUtxoAccountIdByAddress(accountAddress, accountIdsWithoutEvms, chainId)
    }

    // For other non-EVM and non-UTXO chains, find exact chainId + address match
    const accountId = accountIdsWithoutEvms.find(accountId => {
      const { account, chainId: accountChainId } = fromAccountId(accountId)
      return accountChainId === chainId && account.toLowerCase() === normalizedAddress
    })
    return accountId ?? null
  },
)

/**
 * Gets the external address book entry for a given address and chainId
 * Leverages selectAddressBookEntriesByChainId which handles EVM namespace matching
 */
export const selectExternalAddressBookEntryByAddress = createDeepEqualOutputSelector(
  selectAccountAddressParamFromFilter,
  selectChainIdParamFromFilter,
  selectAddressBookEntriesByChainId,
  (address, _, entries) => {
    if (!address) return null
    if (!entries || entries.length === 0) return null

    const normalizedAddress = address.toLowerCase()
    return (
      entries.find(
        entry => entry.isExternal && entry.address.toLowerCase() === normalizedAddress,
      ) ?? null
    )
  },
)
