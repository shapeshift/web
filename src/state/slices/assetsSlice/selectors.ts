import { createSelector } from '@reduxjs/toolkit'
import type { AssetId, ChainId } from '@shapeshiftoss/caip'
import { fromAssetId, isNft } from '@shapeshiftoss/caip'
import { matchSorter } from 'match-sorter'
import createCachedSelector from 're-reselect'
import type { Asset } from 'lib/asset-service'
import type { ReduxState } from 'state/reducer'
import { createDeepEqualOutputSelector } from 'state/selector-utils'
import { selectAssetIdParamFromFilter, selectSearchQueryFromFilter } from 'state/selectors'

import { selectCryptoMarketDataIds } from '../marketDataSlice/selectors'
import { assetIdToFeeAssetId } from '../portfolioSlice/utils'
import { getFeeAssetByAssetId, getFeeAssetByChainId } from './utils'

export const selectAssetById = createCachedSelector(
  (state: ReduxState) => state.assets.byId,
  (_state: ReduxState, assetId: AssetId) => assetId,
  (byId, assetId) => byId[assetId] || undefined,
)((_state: ReduxState, assetId: AssetId | undefined): AssetId => assetId ?? 'undefined')

export const selectAssetByFilter = createCachedSelector(
  (state: ReduxState) => state.assets.byId,
  selectAssetIdParamFromFilter,
  (byId, assetId) => byId[assetId ?? ''] || undefined,
)((_s: ReduxState, filter) => filter?.assetId ?? 'assetId')

export const selectAssetNameById = createSelector(
  selectAssetById,
  (asset): string => asset?.name ?? '',
)

export const selectAssets = createDeepEqualOutputSelector(
  (state: ReduxState) => state.assets.byId,
  byId => byId,
)
export const selectAssetIds = (state: ReduxState) => state.assets.ids

export const selectNftAssetIds = createDeepEqualOutputSelector(
  selectAssetIds,
  (assetIds): AssetId[] => assetIds.filter(assetId => isNft(assetId)),
)

export const selectAssetsByMarketCap = createDeepEqualOutputSelector(
  selectCryptoMarketDataIds,
  selectAssets,
  (marketDataAssetIds, assets): Asset[] => {
    const sortedAssets = marketDataAssetIds.reduce<Asset[]>((acc, assetId) => {
      const asset = assets[assetId]
      if (asset) acc.push(asset)
      return acc
    }, [])

    return sortedAssets
  },
)

export const selectChainIdsByMarketCap = createDeepEqualOutputSelector(
  selectAssetsByMarketCap,
  (sortedAssets: Asset[]): ChainId[] =>
    sortedAssets.reduce<ChainId[]>((acc, { assetId }) => {
      const feeAssetId = assetIdToFeeAssetId(assetId)
      if (feeAssetId !== assetId) return acc
      const { chainId } = fromAssetId(feeAssetId)
      if (!acc.includes(chainId)) acc.push(chainId)
      return acc
    }, []),
)

export const selectFeeAssetByChainId = createCachedSelector(
  selectAssets,
  (_state: ReduxState, chainId: ChainId) => chainId,
  (assetsById, chainId): Asset | undefined => getFeeAssetByChainId(assetsById, chainId),
)((_state: ReduxState, chainId) => chainId ?? 'chainId')

export const selectFeeAssetById = createCachedSelector(
  selectAssets,
  (_state: ReduxState, assetId: AssetId) => assetId,
  (assetsById, assetId): Asset | undefined => getFeeAssetByAssetId(assetsById, assetId),
)((_s: ReduxState, assetId: AssetId) => assetId ?? 'assetId') as (
  state: ReduxState,
  assetId: AssetId,
) => ReturnType<typeof getFeeAssetByAssetId>

export const selectAssetsBySearchQuery = createDeepEqualOutputSelector(
  selectAssetsByMarketCap,
  selectSearchQueryFromFilter,
  (sortedAssets: Asset[], searchQuery?: string): Asset[] => {
    if (!searchQuery) return sortedAssets
    return matchSorter(sortedAssets, searchQuery ?? '', {
      keys: ['name', 'symbol', 'assetId'],
      threshold: matchSorter.rankings.MATCHES,
    })
  },
)
