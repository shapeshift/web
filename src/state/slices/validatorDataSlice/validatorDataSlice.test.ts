import { MOCK_VALIDATORS, SHAPESHIFT_OPPORTUNITY } from 'test/mocks/validators'
import { clearState, store } from 'state/store'

import { SHAPESHIFT_VALIDATOR_ADDRESS } from './const'
import { selectValidatorByAddress } from './selectors'
import { validatorData } from './validatorDataSlice'

describe('validatorDataSlice', () => {
  beforeEach(() => {
    clearState()
  })

  it('returns uninitialized properties for initialState', async () => {
    expect(store.getState().validatorData).toEqual({
      byValidator: {},
      validatorIds: [],
    })
  })

  describe('reducers', () => {
    describe('upsertValidatorData', () => {
      it('updates or inserts validator data in state', async () => {
        store.dispatch(
          validatorData.actions.upsertValidatorData({
            validators: MOCK_VALIDATORS,
          }),
        )
        expect(store.getState().validatorData.byValidator).toMatchSnapshot()
      })
    })

    describe('clear', () => {
      it('clears state back to initial state', async () => {
        store.dispatch(
          validatorData.actions.upsertValidatorData({
            validators: MOCK_VALIDATORS,
          }),
        )

        expect(
          store.getState().validatorData.byValidator[SHAPESHIFT_VALIDATOR_ADDRESS],
        ).toMatchSnapshot()
        expect(store.getState().validatorData.byValidator).toMatchSnapshot()
        store.dispatch(validatorData.actions.clear())

        expect(store.getState().validatorData).toEqual({
          byValidator: {},
          validatorIds: [],
        })
      })
    })
  })

  describe('selectors', () => {
    describe('selectValidatorByAddress', () => {
      it('returns null on initial state', async () => {
        const selected = selectValidatorByAddress(store.getState(), SHAPESHIFT_VALIDATOR_ADDRESS)
        expect(selected).toBeNull()
      })

      describe('validators are loaded', () => {
        beforeEach(() => {
          store.dispatch(
            validatorData.actions.upsertValidatorData({
              validators: MOCK_VALIDATORS,
            }),
          )
        })

        it('returns null when validator data is not present in state', async () => {
          store.dispatch(
            validatorData.actions.upsertValidatorData({
              validators: MOCK_VALIDATORS,
            }),
          )

          const selected = selectValidatorByAddress(
            store.getState(),
            'cosmosvaloper1xxjvzwjpsf6ktuffkcpx29ut9qfrn0my8xdtqd',
          )
          expect(selected).toBeNull()
        })

        it('returns validator info when validator data is present in state', async () => {
          store.dispatch(
            validatorData.actions.upsertValidatorData({
              validators: MOCK_VALIDATORS,
            }),
          )

          const selected = selectValidatorByAddress(store.getState(), SHAPESHIFT_VALIDATOR_ADDRESS)
          expect(selected).toEqual(SHAPESHIFT_OPPORTUNITY)
        })
      })
    })
  })
})
