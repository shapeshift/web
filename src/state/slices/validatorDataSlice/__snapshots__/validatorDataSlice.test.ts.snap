// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validatorDataSlice reducers clear clears state back to initial state 1`] = `
Object {
  "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
  "apr": "0.24",
  "commission": "0.100000000000000000",
  "moniker": "Shapeshift DAO",
  "tokens": "111111",
}
`;

exports[`validatorDataSlice reducers clear clears state back to initial state 2`] = `
Object {
  "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf": Object {
    "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
    "apr": "0.24",
    "commission": "0.100000000000000000",
    "moniker": "Shapeshift DAO",
    "tokens": "111111",
  },
  "cosmosvaloper19f0w9svr905fhefusyx4z8sf83j6et0g9l5yhl": Object {
    "address": "cosmosvaloper19f0w9svr905fhefusyx4z8sf83j6et0g9l5yhl",
    "apr": "0.1682133662",
    "commission": "0.010000000000000000",
    "moniker": "NodeStake.top",
    "tokens": "1366570093",
  },
  "cosmosvaloper1xhhquwctvwm40qn3nmmqq067pc2gw22eqkwgt0": Object {
    "address": "cosmosvaloper1xhhquwctvwm40qn3nmmqq067pc2gw22eqkwgt0",
    "apr": "0.1682133662",
    "commission": "0.010000000000000000",
    "moniker": "stake2earn",
    "tokens": "4474530413",
  },
}
`;

exports[`validatorDataSlice reducers upsertValidatorData updates or inserts validator data in state 1`] = `
Object {
  "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf": Object {
    "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
    "apr": "0.24",
    "commission": "0.100000000000000000",
    "moniker": "Shapeshift DAO",
    "tokens": "111111",
  },
  "cosmosvaloper19f0w9svr905fhefusyx4z8sf83j6et0g9l5yhl": Object {
    "address": "cosmosvaloper19f0w9svr905fhefusyx4z8sf83j6et0g9l5yhl",
    "apr": "0.1682133662",
    "commission": "0.010000000000000000",
    "moniker": "NodeStake.top",
    "tokens": "1366570093",
  },
  "cosmosvaloper1xhhquwctvwm40qn3nmmqq067pc2gw22eqkwgt0": Object {
    "address": "cosmosvaloper1xhhquwctvwm40qn3nmmqq067pc2gw22eqkwgt0",
    "apr": "0.1682133662",
    "commission": "0.010000000000000000",
    "moniker": "stake2earn",
    "tokens": "4474530413",
  },
}
`;
